File: src/character.ts
import { Counter } from './counter';
import { HasHp } from './hasHp';
import { InitiativeItem } from './initiativeItem';
export class Character implements InitiativeItem, HasHp {
    private _name: string;
    public get name(): string {
        return this._name;
    }
    public set name(value: string) {
        this._name = value;
    }

    private _initiative: number;
    public get initiative(): number {
        return this._initiative;
    }
    public set initiative(value: number) {
        this._initiative = value;
    }

    private _conditions: string[];
    public get conditions(): string[] {
        return this._conditions;
    }
    private set conditions(value: string[]) {
        this._conditions = value;
    }

    public get hp(): number {
        return this.getCounter('hp');
    }
    public set hp(value: number) {
        this.setCounter('hp', value);
    }

    public get maxHp(): number {
        return this.getCounterMax('hp');
    }
    public set maxHp(value: number) {
        // if (this.maxHp === null) {
        //     this._maxHp = value;
        //     this.hp = this.hp;
        // } else {
        //     const hpMaxDiff = value - this.maxHp;
        //     this._maxHp = value;
        //     // when hpMax changes, hp goes up or down with it
        //     this.heal(hpMaxDiff);
        // }
        let hpMaxDiff = value - this.maxHp;
        this.setCounterMax('hp', value);
        this.increaseCounter('hp', hpMaxDiff);
    }

    private _tempHp: number;
    public get tempHp(): number {
        return this._tempHp;
    }
    private set tempHp(value: number) {
        this._tempHp = value;
        if (this.tempHp != null && this.tempHp <= 0) {
            this._tempHp = null
        }
    }

    private _counters: Counter[];
    public get counters(): Counter[] {
        return this._counters;
    }
    public set counters(value: Counter[]) {
        this._counters = value;
    }

    constructor(
        name: string,
        initiative: number = null,
        conditions: string[] = [],
        hp: number = null,
        maxHp: number = null,
        tempHp: number = null
    ) {
        this.name = name;
        this.initiative = initiative;
        this.conditions = conditions;
        this.tempHp = tempHp;

        this.counters = [];

        this.addCounter('hp', hp, maxHp);
    }

    heal(healAmount: number) {
        if (this.hp !== null) {
            this.hp+=healAmount;
        }
    }

    getInitiativeString() {
        let toReturn = `\t${this.name}: ${this.initiative}`;
        if (this.hp != null && !isNaN(this.hp)) {
            toReturn+='\n\t\t';
            if (this.tempHp != null && !isNaN(this.tempHp)) {
                toReturn += `${this.tempHp} temp hp -- `;
            }
            toReturn += `${this.hp}`
            if (this.maxHp != null && !isNaN(this.maxHp)) {
                toReturn += `/${this.maxHp}`;
            }
            toReturn += ' hp';
        }

        if (this.counters.length !== 0) {
            this.counters.forEach((counter) => {
                if (counter.name != 'hp') {
                    toReturn += `\n\t\t\t${counter.name}: ${counter.value}`;
                    if (counter.max != null) {
                        toReturn += `/${counter.max}`;
                    }
                }
            })
        }

        if (this.conditions.length !== 0) {
            toReturn += `\n\t\tConditions: ${this.conditions[0]}`;
            for (let i = 1; i < this.conditions.length; i++) {
                toReturn += `, ${this.conditions[i]}`;
            }
        }

        toReturn += '\n';

        return toReturn;
    }

    damage(damageAmount: number) {
        if (this.tempHp !== null) {
            const leftOverDamage = damageAmount - this.tempHp;
            this.tempHp-=damageAmount;
            if (leftOverDamage > 0) {
                this.hp -= leftOverDamage;
            }
        } else {
            this.hp -= damageAmount;
        }

        if (this.hp <= 0) {
            this.addConditions(['incapacitated']);
        }
    }

    addConditions(newConditions: string[]) {
        newConditions.forEach((condition)=>{
            this.conditions.push(condition.trim());
        })
    }

    removeConditions(conditions: string[]) {
        conditions.forEach((toRemove)=>{
            this.conditions = this.conditions.filter(
                (value: string) => value !== toRemove
            );
        })
    }

    removeCounter(counterName: string) {
        this.counters = this.counters.filter(
            (counter: Counter) => counter.name !== counterName
        );
    }

    addCounter(name: string, value: number, max: number) {
        if (this.getCounter(name) == null) {
            this.counters.push(new Counter(
                name,
                value,
                max
            ));
        }
    }

    private _getCounter(name: string): Counter {
        let toReturn: Counter = null;
        this.counters.forEach((counter) => {
            if (counter.name == name) {
                toReturn = counter;
            }
        })
        return toReturn;
    }

    getCounter(name: string): number {
        let myCounter = this._getCounter(name);
        if (myCounter == null) {
            return null;
        } else {
            return myCounter.value;
        }
    }

    getCounterMax(name: string): number {
        let myCounter = this._getCounter(name);
        if (myCounter == null) {
            return null;
        } else {
            return myCounter.max;
        }
    }

    setCounter(name: string, value: number) {
        let myCounter = this._getCounter(name);
        if (myCounter != null) {
            myCounter.value = value;
        }
    }

    setCounterMax(name: string, max: number) {
        let myCounter = this._getCounter(name);
        if (myCounter != null) {
            myCounter.max = max;
        }
    }

    increaseCounter(name: string, delta: number) {
        let myCounter = this._getCounter(name);
        if (myCounter != null) {
            myCounter.value += delta;
        }
    }

    increaseCounterMax(name: string, delta: number) {
        let myCounter = this._getCounter(name);
        if (myCounter != null) {
            myCounter.max += delta;
        }
    };
}

File: src/trash/python/character.py
class Character(object):
    def __init__(self, name, initiative, conditions, hp=None, maxHp=None, tempHp=None):
        try:
            self.name = name
            self.initiative = int(initiative)

            if maxHp is not None:
                self.maxHp = int(maxHp)
            else:
                self.maxHp = None
            
            if hp is not None:
                self.hp = int(hp)
                self.setHp(self.hp)
            else:
                self.hp = None    

            if tempHp is not None:
                self.tempHp = int(tempHp)
            else:
                self.tempHp = None   
            
            self.conditions = conditions
        except:
            print('oopsy woopsy! looks like you tried to initialize a character with bad inputs UwU')
            self.initiative = 1
            self.maxHp = None
            self.hp = None
            self.conditions = []

    def setHp(self, newHp):
        if self.maxHp is None or newHp is None:
            self.hp = newHp
            return

        if newHp > self.maxHp:
            self.hp = self.maxHp
        else:
            self.hp = newHp

    def setMaxHp(self, newHpMax):
        if self.maxHp is None:
            self.maxHp = newHpMax
            self.setHp(self.hp)
        else:
            hpMaxDiff = newHpMax - self.maxHp
            self.maxHp = newHpMax
            # when hpMax changes, hp goes up or down with it
            self.heal(hpMaxDiff)
    
    def setTempHp(self, newTempHp):
        self.tempHp = newTempHp
        if self.tempHp <= 0:
            self.tempHp = None

    def heal(self, healAmount):
        if self.hp is not None:
            self.setHp(self.hp+healAmount)

    def getInitiativeString(self):
        toReturn = '\t%s: %d' % (self.name, self.initiative)
        if self.hp is not None:
            toReturn+='\n\t\t'
            if self.tempHp is not None:
                toReturn += '%d temp hp -- ' % self.tempHp
            toReturn += '%d' % self.hp
            if self.maxHp is not None:
                toReturn += '/%d' % self.maxHp
            toReturn += ' hp'
        if self.conditions.__len__() != 0:
            toReturn+= '\n\t\tConditions: %s' % self.conditions[0]
            for condition in self.conditions[1 : self.conditions.__len__()]:
                toReturn += ', %s' % condition
        toReturn += '\n'
        
        return toReturn

    def damage(self, damageAmount):
        if self.tempHp is not None:
            leftOverDamage = damageAmount - self.tempHp
            self.setTempHp(self.tempHp - damageAmount)
            if leftOverDamage > 0:
                self.setHp(self.hp-leftOverDamage)
        else:
            self.setHp(self.hp-damageAmount)
        
        if self.hp <= 0:
            self.addConditions(['incapacitated'])

    def addConditions(self, newConditions):
        for condition in newConditions:
            self.conditions.append(condition)
        self.apple = 'poop'

    def removeConditions(self, conditions):
        for condition in conditions:
            i = 0
            while i < self.conditions.__len__():
                if self.conditions[i].lower() == condition:
                    self.conditions.pop(i)
                i+=1